scala/integration/intellij-bazel/src/org/jetbrains/plugins/scala/bazel/BazelTestRunLineMarkerContributor.scala

```diff
@@ -0,0 +1,77 @@
+package org.jetbrains.plugins.scala.bazel
+
+import com.intellij.psi.impl.source.tree.LeafPsiElement
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.execution.PsiLocation
+
+import org.jetbrains.bazel.ui.gutters.BazelRunLineMarkerContributor
+import org.jetbrains.plugins.scala.lang.lexer.ScalaTokenTypes
+import org.jetbrains.plugins.scala.lang.psi.api.expr.{ScInfixExpr, ScMethodCall, ScReferenceExpression}
+import org.jetbrains.plugins.scala.lang.psi.api.statements.ScFunctionDefinition
+import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.{ScClass, ScTypeDefinition}
+
+import org.jetbrains.plugins.scala.extensions.ObjectExt
+import org.jetbrains.plugins.scala.testingSupport.test.scalatest.ScalaTestConfigurationProducer
+
+import java.util
+
+class BazelTestRunLineMarkerContributor extends BazelRunLineMarkerContributor {
+  private val TEST_ARG = "-t"
+
+  override def isDumbAware: Boolean = true
+
+  override def shouldAddMarker(psiElement: PsiElement): Boolean =
+    psiElement match {
+      case leaf: LeafPsiElement if leaf.getElementType == ScalaTokenTypes.tIDENTIFIER =>
+        leaf.getParent match {
+          case _: ScTypeDefinition | _: ScReferenceExpression => true
+          case _ => false
+        }
+      case _ =>
+        false
+    }
+
+  override def getSingleTestFilter(element: PsiElement): String =
```

Comment:
Hm, so the `getSingleTestFilter` method comes from `BazelJavaRunLineMarkerContributor` (and from `BazelRunLineMarkerContributor`) and has some default implementation.

It looks then, that it's implied that one must not implement 2 separate providers for tests and non-tests (It's hard to tell what was the original idea without the documentation).

However, currently 2 providers are implemented:
`BazelScalaRunLineMarkerContributor` and `BazelTestRunLineMarkerContributor` which means that 2 test logics will be executed, which seems "off".

cc @jastice

Reply:
You're right, I see the concern now. When I implemented this, I followed the pattern from the existing ScalaRunLineMarkerContributor and ScalaTestRunLineMarkerProvider, which are split into separate classes. I tried to keep my changes less invasive by maintaining this same structure.

However, I agree that the presence of the getSingleTestFilter method in BazelRunLineMarkerContributor does suggest that both test and non-test execution should be implemented in a single class. 
Should I try to merge BazelScalaRunLineMarkerContributor and BazelTestRunLineMarkerContributor into a single contributor class?

Reply:
Yes, please merge them given we are talking about maintaining code in Scala Plugin repository now.
We still could extract test logic calculation to separate some separate class to not to mix test and non-test logic, especially if  there is a lot of code in the merged version.

The main point is to not register 2 separate providers

Reply:
Got it, thanks for the clarity

Reply:
@unkarjedy Gentle ping on this PR 🙂

Reply:
Actually the main point was to not register a single instance of `BazelJavaRunLineMarkerContributor` (given that this is what "Bazel" plugin dictates.
But it sill makes sense to extract test-related logic to a different file/class/etc...
This will simplify the maintenance.

I will do it myself while merging the PR